### PR TITLE
executor: fix wrong result when select `NULL` value on index column

### LINKED ISSUE #16921: executor: select `NULL` value on index column
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. What did you do?
```sql
mysql> create table t (a float);                            
mysql> create index a on t(a);                              
mysql> insert into t values (1.0), (NULL);                  
```

### 2. What did you expect to see?
With MySQL 8.0.19:
```sql
mysql> select `a` from `t` use index (a) where !`a`;                                                                 
+---+
| a |
+---+
mysql> select `a` from `t` ignore index (a) where !`a`;                                                              
+---+
| a |
+---+
mysql> select `a` from `t` use index (a) where `a`;                                                                  
+-----+
| a   |
+-----+
| 1.0 |
+-----+
mysql> select `a` from `t` ignore index (a) where `a`;                                                               
+-----+
| a   |
+-----+
| 1.0 |
+-----+
```

### 3. What did you see instead?
With TiDB (and mocktikv/TiKV):
```sql
mysql> select `a` from `t` use index (a) where !`a`;                                                                                              
+--------+
| a      |
+--------+
| <null> |
+--------+
mysql> select `a` from `t` ignore index (a) where !`a`;                                                                                           
+---+
| a |
+---+
mysql> select `a` from `t` use index (a) where `a`;                                                                                               
+-----+
| a   |
+-----+
| 1.0 |
+-----+
mysql> select `a` from `t` ignore index (a) where `a`;                                                                                            
+-----+
| a   |
+-----+
| 1.0 |
+-----+
```

### 4. What version of TiDB are you using? (`tidb-server -V` or run `select tidb_version();` on TiDB)
master

### PR DESCRIPTION
**Pull request by Reminiscent**
### What problem does this PR solve?

Related to [PR#16108](https://github.com/pingcap/tidb/pull/16108)
Issue Number: close #16921 

Problem Summary:
When we push down the expression `not`, we shouldn't warp it with expression `isTrue`.

### What is changed and how it works?
Do not warp expression `not` with expression `isTrue` when `not` is pushed down.

### Check List 

Tests 

- Unit test
- Integration test

### Release note 
- fix wrong result when select `NULL` value on index column

### CODE REVIEW

**Comment by Reminiscent** [comment]
In mysql8.0
```
mysql> create table t (a float);
Query OK, 0 rows affected (0.06 sec)

mysql> create index a on t(a);
Query OK, 0 rows affected (0.03 sec)
Records: 0  Duplicates: 0  Warnings: 0

mysql> insert into t values (1.0), (NULL);
Query OK, 2 rows affected (0.01 sec)
Records: 2  Duplicates: 0  Warnings: 0

mysql> select `a` from `t` use index (a) where not not `a`;
+------+
| a    |
+------+
|    1 |
+------+
1 row in set (0.00 sec)

mysql> select `a` from `t` use index (a) where !!`a`;
Empty set, 1 warning (0.00 sec)

mysql> show warnings;
+---------+------+-----------------------------------------------------------------------------------+
| Level   | Code | Message                                                                           |
+---------+------+-----------------------------------------------------------------------------------+
| Warning | 1287 | '!' is deprecated and will be removed in a future release. Please use NOT instead |
+---------+------+-----------------------------------------------------------------------------------+
1 row in set (0.00 sec)
```

**Comment by Reminiscent** [comment]
After this PR, TiDB master
```
mysql> create table t (a float);                            
Query OK, 0 rows affected (0.01 sec)

mysql> create index a on t(a);                              
Query OK, 0 rows affected (0.03 sec)

mysql> insert into t values (1.0), (NULL); 
Query OK, 2 rows affected (0.01 sec)
Records: 2  Duplicates: 0  Warnings: 0

mysql> 
mysql> select `a` from `t` use index (a) where not not `a`; 
+------+
| a    |
+------+
|    1 |
+------+
1 row in set (0.00 sec)

mysql> select `a` from `t` use index (a) where !!`a`;  
+------+
| a    |
+------+
|    1 |
+------+
1 row in set (0.00 sec)
```

**Comment by XuHuaiyu** [comment]
CI failed
```
[2020-05-12T11:55:57.340Z] FAIL: physical_plan_test.go:368: testPlanSuite.TestRefine

[2020-05-12T11:55:57.340Z] 

[2020-05-12T11:55:57.340Z] physical_plan_test.go:399:

[2020-05-12T11:55:57.340Z]     c.Assert(core.ToString(p), Equals, output[i].Best, comment)

[2020-05-12T11:55:57.340Z] ... obtained string = "IndexReader(Index(t.f)[[NULL,+inf]]->Sel([not(test.t.a)]))"

[2020-05-12T11:55:57.340Z] ... expected string = "PointGet(Handle(t.a)0)"

[2020-05-12T11:55:57.340Z] ... for select a from t where not a
```

**Comment by Reminiscent** [comment]
@wshwsh12 @XuHuaiyu PTAL

**Comment by Reminiscent** [comment]
> ```
> func (r *builder) buildFromColumn(expr *expression.Column) []point {
> 	// column name expression is equivalent to column name is true.
> 	startPoint1 := point{value: types.MinNotNullDatum(), start: true}
> 	endPoint1 := point{excl: true}
> 	endPoint1.value.SetInt64(0)
> 	startPoint2 := point{excl: true, start: true}
> 	startPoint2.value.SetInt64(0)
> 	endPoint2 := point{value: types.MaxValueDatum()}
> 	return []point{startPoint1, endPoint1, startPoint2, endPoint2}
> }
> ```
> 
> `Column#i` is treated the same with `Column#i is true` when building range.
> 
> **UPD** We may need to change this one directly

```
func (r *builder) buildFromIsTrue(expr *expression.ScalarFunction, isNot int) []point {
	if isNot == 1 {
		// NOT TRUE range is {[null null] [0, 0]}
		startPoint1 := point{start: true}
		endPoint1 := point{}
		startPoint2 := point{start: true}
		startPoint2.value.SetInt64(0)
		endPoint2 := point{}
		endPoint2.value.SetInt64(0)
		return []point{startPoint1, endPoint1, startPoint2, endPoint2}
	}
	// TRUE range is {[-inf 0) (0 +inf]}
	startPoint1 := point{value: types.MinNotNullDatum(), start: true}
	endPoint1 := point{excl: true}
	endPoint1.value.SetInt64(0)
	startPoint2 := point{excl: true, start: true}
	startPoint2.value.SetInt64(0)
	endPoint2 := point{value: types.MaxValueDatum()}
	return []point{startPoint1, endPoint1, startPoint2, endPoint2}
}
```
Maybe we should the change this one. We need to construct a range based on the expression's variable `isTrue.keepNull`. But it seems not easy to get the `keepNull` variable. When the condition `isTrue.keepNull = true` , is it possible for the `buildFromIsTrue` function to be called?

File: planner/core/testdata/plan_suite_out.json [hunk 1]
```diff
@@ -939,7 +939,7 @@
       },
       {
         "SQL": "select a from t where not a",
-        "Best": "PointGet(Handle(t.a)0)"
+        "Best": "IndexReader(Index(t.f)[[NULL,+inf]]->Sel([not(test.t.a)]))"
```
Comment (Reminiscent) [inline planner/core/testdata/plan_suite_out.json:942]: Related to https://github.com/pingcap/tidb/pull/1828.

**Comment by Reminiscent** [comment]
@XuHuaiyu @winoros @wshwsh12 PTAL

File: executor/executor_test.go [hunk 2]
```diff
@@ -5524,6 +5524,24 @@ func (s *testSuite1) TestIssue16854(c *C) {
...
+
+	tk.MustExec("drop table if exists t;")
+	tk.MustExec("create table t (a float);")
+	tk.MustExec("create index a on t(a);")
+	tk.MustExec("insert into t values (1.0), (NULL), (0), (2.0);")
+	tk.MustQuery("select `a` from `t` use index (a) where !`a`;").Check(testkit.Rows("0"))
+	tk.MustQuery("select `a` from `t` ignore index (a) where !`a`;").Check(testkit.Rows("0"))
+	tk.MustQuery("select `a` from `t` use index (a) where `a`;").Check(testkit.Rows("1", "2"))
+	tk.MustQuery("select `a` from `t` ignore index (a) where `a`;").Check(testkit.Rows("1", "2"))
+	tk.MustQuery("select a from t use index (a) where not a is true;").Check(testkit.Rows("<nil>", "0"))
```
Comment (Reminiscent) [inline executor/executor_test.go:5538]: The answer of `not a is true`  is different from `not a`. So we shouldn't wrap `not` with `isTrue`.

**Comment by Reminiscent** [comment]
This PR can solve all problems encountered now. But I don't know if there may be other problems due to the [comment](https://github.com/pingcap/tidb/pull/17117#issuecomment-628984320).

**Comment by SunRunAway** [comment]
> After this PR, TiDB master
> 
> ```
> mysql> create table t (a float);                            
> Query OK, 0 rows affected (0.01 sec)
> 
> mysql> create index a on t(a);                              
> Query OK, 0 rows affected (0.03 sec)
> 
> mysql> insert into t values (1.0), (NULL); 
> Query OK, 2 rows affected (0.01 sec)
> Records: 2  Duplicates: 0  Warnings: 0
> 
> mysql> 
> mysql> select `a` from `t` use index (a) where not not `a`; 
> +------+
> | a    |
> +------+
> |    1 |
> +------+
> 1 row in set (0.00 sec)
> 
> mysql> select `a` from `t` use index (a) where !!`a`;  
> +------+
> | a    |
> +------+
> |    1 |
> +------+
> 1 row in set (0.00 sec)
> ```

It's a known bug in mysql, https://bugs.mysql.com/bug.php?id=55477

**Comment by Reminiscent** [comment]
@XuHuaiyu @winoros @wshwsh12 Friendly ping. PTAL

**Comment by wshwsh12** [comment]
@Reminiscent  This pr leads to the condition like`not a` can't use index. For example, cc https://github.com/pingcap/tidb/issues/19100 table `t2`.

File: executor/executor_test.go (hunk 2 above)
Comment (wshwsh12) [reply executor/executor_test.go:5538]: Could wrap the column with IsFalse?

**Comment by Reminiscent** [comment]
> @Reminiscent This pr leads to the condition like`not a` can't use index. For example, cc #19100 table `t2`.

OK, I will see it in the weekend.....

**Comment by Reminiscent** [comment]
@wshwsh12 OK, I will see it at the weekend.....

**Comment by wshwsh12** [comment]
I fix the issue in pr https://github.com/pingcap/tidb/pull/19385. 
Thanks for your contribution, and feel free to reopen this pr or create another pr/issue if you any idea else.

**Pull request closed without merging**